### PR TITLE
[Prototype] Goodbye Tree Dump Control, Hello XAML Jest Snapshots 

### PR DESCRIPTION
**Pull request by NickGerleman**
Much NYI in the prototype, but a proof of concept of sending tree dumps over the wire, then using Jest snapshot capability to compare them.

image

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7078)

### CODE REVIEW

File: packages/e2e-test-app/windows/ReactUWPTestApp/TreeDumpControlViewManager.cs [hunk 1]
```diff
@@ -1,365 +0,0 @@
...
-            m_textBlock.LayoutUpdated += async (source, e) =>
-            {
-                await KnownFolders.DocumentsLibrary.CreateFolderAsync("TreeDump", CreationCollisionOption.OpenIfExists);
-                ApplicationView.GetForCurrentView().TryResizeView(new Size(800, 600));
-                var bounds = ApplicationView.GetForCurrentView().VisibleBounds;
-                if (bounds.Width != 800 || bounds.Height != 600)
-                {
-                    // Dump disabled when window size is not 800x600!
-                    UpdateResult(false /*matchDump*/ , $"Window has been resized, dump comparison is only valid at default launch size: 800x600!, current size: {bounds.Width}x{bounds.Height}");
-                }
```
Comment (asklar) [inline packages/e2e-test-app/windows/ReactUWPTestApp/TreeDumpControlViewManager.cs:42]: this logic is still relevant, the dumps won't match at different resolutions

File: packages/node-rnw-rpc/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+### Alternative approaches
+- **Named pipe created inside AppContainer:** Node provides first-class support for named-pipes through `net.socket`. It is possible to create a named pipe from a UWP process which a Win32 process can then connect to, allowing traditional server/client roles. Doing this requires locating the resource local to the AppContainer, using an API like `GetAppContainerNamedObjectPath`. This likely requires FFI (with node-gyp) or native extenssions.
+- **Proxy to full-trust process** The RNW application could create a full-trust process which proxie from TCP to named-pipe internal to the AppContainer. This bypasses inbound loopback restrictions, but adds complexity of deploying a separate process, requires two IPC channels.
+
+### Protocol
+node-rnw-ipc uses a TCP channel, sending JSON messages across the wire prefixed with length. Messages themselves are conformant to the [JSON-RPC](https://en.wikipedia.org/wiki/JSON-RPC) version 2.0 protocol.
```
Comment (asklar) [inline packages/node-rnw-rpc/README.md:56]: shouldn't this say rpc instead of ipc?

File: packages/node-rnw-rpc/windows/NodeRpc/Server.idl [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (asklar) [inline packages/node-rnw-rpc/windows/NodeRpc/Server.idl:1]: are these going to be public APIs? if so can you add the comments as doc_string so we can generate docs from it?

File: packages/@react-native-windows/tester/src/js/examples-win/LegacyTests/TextBackgroundColorTestPage.tsx [hunk 4]
```diff
@@ -60,16 +58,6 @@ export function TextBackgroundColorTestPage() {
...
 
-      <TreeDumpControl
-        style={{
-          height: 150,
-          width: 500,
-          margin: 10,
-        }}
-        dumpID={'TextColorBackground'}
-        uiaID={'text-color-background-view'}
-        additionalProperties={['TextHighlighters']}
```
Comment (asklar) [inline packages/@react-native-windows/tester/src/js/examples-win/LegacyTests/TextBackgroundColorTestPage.tsx:71]: how are you going to ask for this and other additional properties to the tree dumper?

File: packages/e2e-test-app/windows/ReactUWPTestApp/MainPage.xaml.cs [hunk 5]
```diff
@@ -16,6 +22,30 @@ public MainPage()
             var app = Application.Current as App;
             myRootView.ReactNativeHost = app.Host;
             myRootView.ComponentName = "RNTesterApp";
+
+
+            var handler = new Handler();
+            handler.BindOperation("DumpVisualTree", (payload) =>
+            {
+                var rootDump = VisualTreeDumper.DumpTree(this, null, new string[] { }, DumpTreeMode.Json);
```
Comment (asklar) [inline packages/e2e-test-app/windows/ReactUWPTestApp/MainPage.xaml.cs:30]: looks like this is missing the additional properties array?

File: packages/e2e-test-app/windows/ReactUWPTestApp/MainPage.xaml.cs [hunk 6]
```diff
@@ -16,6 +22,30 @@ public MainPage()
...
+
+            var server = new Server(handler);
+            var tsk = LoopServer(server);
         }
+
+        async Task LoopServer(Server server)
+        {
+            while (true)
+            {
+                await server.ProcessAllClientRequests(8603, TimeSpan.FromMilliseconds(50));
```
Comment (asklar) [inline packages/e2e-test-app/windows/ReactUWPTestApp/MainPage.xaml.cs:46]: does the port need to be made configurable ?

File: packages/e2e-test-app/windows/ReactUWPTestApp/TreeDumpControlViewManager.cs [hunk 7]
```diff
@@ -1,365 +0,0 @@
...
-                new Run () { Text = " does not match " },
-                masterLink
-            };
-
-            #region Diff support - Replace with LaunchUriAsync when we find the VSCode protocol handler Uri for diffing
-            string code_cmd = Environment.ExpandEnvironmentVariables(@"%UserProfile%\AppData\Local\Programs\Microsoft VS Code\bin\code.cmd");
-            Hyperlink diffLink = new Hyperlink();
-            diffLink.Click += (_1, _2) =>
-            {
-                string commandLine = $"code.cmd --diff \"{masterFile.Path}\" \"{outFile.Path}\"";
```
Comment (asklar) [inline packages/e2e-test-app/windows/ReactUWPTestApp/TreeDumpControlViewManager.cs:260]: being able to click in the UI to open the difference is kind of a nice feature... can this kind of functionality be preserved?

File: .ado/jobs/integration-test.yml [hunk 8]
```diff
@@ -57,11 +57,6 @@ jobs:
         displayName: yarn windows --no-launch
```
Comment (asklar) [inline .ado/jobs/integration-test.yml:57]: you'll want to revise the e2e test yml file since it's integrated with uploading the masters and the outputs

**Review by asklar — CHANGES_REQUESTED** [review]

**Comment by NickGerleman** [comment]
@asklar, this is a rough prototype not ready for review/merge. Goal here is a POC of overall structure.

File: packages/e2e-test-app/windows/ReactUWPTestApp/TreeDumpControlViewManager.cs (hunk 1 above)
Comment (NickGerleman) [reply packages/e2e-test-app/windows/ReactUWPTestApp/TreeDumpControlViewManager.cs:42]: The root view in the window has a fixed size.

File: packages/node-rnw-rpc/README.md (hunk 2 above)
Comment (NickGerleman) [reply packages/node-rnw-rpc/README.md:56]: RPC: Remote Procedure Call
IPC: Inter-Process-Communication

RPC is the action. IPC is the channel.

Comment (asklar) [reply packages/node-rnw-rpc/README.md:56]: yes I'm aware of the acronym. My point is you call the package node-rnw-rpc elsewhere but here you call it node-rnw-ipc.

File: packages/node-rnw-rpc/windows/NodeRpc/Server.idl (hunk 3 above)
Comment (NickGerleman) [reply packages/node-rnw-rpc/windows/NodeRpc/Server.idl:1]: Will not be public as part of RNW docs. Likely just small package README.

File: packages/@react-native-windows/tester/src/js/examples-win/LegacyTests/TextBackgroundColorTestPage.tsx (hunk 4 above)
Comment (NickGerleman) [reply packages/@react-native-windows/tester/src/js/examples-win/LegacyTests/TextBackgroundColorTestPage.tsx:71]: Can send arbitrary data over IPC channel. Will add extra param for it.

File: packages/e2e-test-app/windows/ReactUWPTestApp/TreeDumpControlViewManager.cs (hunk 7 above)
Comment (NickGerleman) [reply packages/e2e-test-app/windows/ReactUWPTestApp/TreeDumpControlViewManager.cs:260]: Jest will show colorized diffs in the test runner, including in CI output for each failing test. See the screenshot in description.

File: packages/node-rnw-rpc/README.md (hunk 2 above)
Comment (NickGerleman) [reply packages/node-rnw-rpc/README.md:56]: Oh, I thought you were pointing to "IPC" on the first highlighted line. Will comb through for typos later.

**Pull request closed without merging**